File: packages/web/src/board.ts
export type Point = { x: number; y: number }
export type View = Point & { zoom: number }
export type DrawingTool = 'pen' | 'rectangle' | 'ellipse' | 'line'
export type Tool = DrawingTool | 'eraser' | 'hand'

// Plain document data, independent of React and the eventual sync provider.
export type BoardElement = {
  id: string
  type: DrawingTool
  color: string
  width: number
  points: Point[]
}

export const MIN_ZOOM = 0.25
export const MAX_ZOOM = 4
export const INITIAL_VIEW: View = { x: 0, y: 0, zoom: 1 }

export function toWorld(point: Point, view: View): Point {
  return { x: (point.x - view.x) / view.zoom, y: (point.y - view.y) / view.zoom }
}

export function zoomAt(view: View, anchor: Point, zoom: number): View {
  const nextZoom = Math.min(MAX_ZOOM, Math.max(MIN_ZOOM, zoom))
  const world = toWorld(anchor, view)
  return { x: anchor.x - world.x * nextZoom, y: anchor.y - world.y * nextZoom, zoom: nextZoom }
}

export function constrainEnd(start: Point, end: Point, type: DrawingTool): Point {
  const dx = end.x - start.x
  const dy = end.y - start.y
  if (type === 'line') {
    const angle = Math.round(Math.atan2(dy, dx) / (Math.PI / 4)) * (Math.PI / 4)
    const length = Math.hypot(dx, dy)
    return { x: start.x + Math.cos(angle) * length, y: start.y + Math.sin(angle) * length }
  }
  const side = Math.max(Math.abs(dx), Math.abs(dy))
  return { x: start.x + (dx < 0 ? -side : side), y: start.y + (dy < 0 ? -side : side) }
}

export function bounds(element: BoardElement) {
  const [start, end = start] = element.points
  return { x: Math.min(start.x, end.x), y: Math.min(start.y, end.y), width: Math.abs(end.x - start.x), height: Math.abs(end.y - start.y) }
}

export function outline(element: BoardElement): Point[] {
  if (element.type === 'pen' || element.type === 'line') return element.points
  const { x, y, width, height } = bounds(element)
  if (element.type === 'rectangle') {
    return [{ x, y }, { x: x + width, y }, { x: x + width, y: y + height }, { x, y: y + height }, { x, y }]
  }
  return Array.from({ length: 97 }, (_, i) => {
    const angle = (i / 96) * Math.PI * 2
    return { x: x + width / 2 + Math.cos(angle) * width / 2, y: y + height / 2 + Math.sin(angle) * height / 2 }
  })
}

function pointSegmentDistance(p: Point, a: Point, b: Point) {
  const dx = b.x - a.x
  const dy = b.y - a.y
  const lengthSquared = dx * dx + dy * dy
  const t = lengthSquared === 0 ? 0 : Math.max(0, Math.min(1, ((p.x - a.x) * dx + (p.y - a.y) * dy) / lengthSquared))
  return Math.hypot(p.x - a.x - t * dx, p.y - a.y - t * dy)
}

function cross(a: Point, b: Point, c: Point) {
  return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x)
}

function segmentDistance(a: Point, b: Point, c: Point, d: Point) {
  // Strict crossing; collinear and endpoint cases are covered by distances.
  if (cross(a, b, c) * cross(a, b, d) < 0 && cross(c, d, a) * cross(c, d, b) < 0) return 0
  return Math.min(pointSegmentDistance(a, c, d), pointSegmentDistance(b, c, d), pointSegmentDistance(c, a, b), pointSegmentDistance(d, a, b))
}

export function hitsElement(element: BoardElement, from: Point, to: Point, radius: number) {
  const points = outline(element)
  const threshold = radius + element.width / 2
  if (points.length === 1) return pointSegmentDistance(points[0], from, to) <= threshold
  return points.some((point, i) => i > 0 && segmentDistance(from, to, points[i - 1], point) <= threshold)
}

File: packages/native/src/commonMain/kotlin/meshboard/Board.kt
package meshboard

import kotlin.math.*

data class Point(val x: Double, val y: Double)
data class View(val x: Double = 0.0, val y: Double = 0.0, val zoom: Double = 1.0) {
    fun world(p: Point) = Point((p.x - x) / zoom, (p.y - y) / zoom)
    fun zoomAt(anchor: Point, requested: Double): View {
        val z = requested.coerceIn(.25, 4.0)
        val p = world(anchor)
        return View(anchor.x - p.x * z, anchor.y - p.y * z, z)
    }
}
enum class Tool(val wire: String, val label: String, val shortcut: String) {
    Pen("pen", "Pen", "P"), Eraser("eraser", "Eraser", "E"),
    Rectangle("rectangle", "Rectangle", "R"), Ellipse("ellipse", "Ellipse", "O"),
    Line("line", "Line", "L"), Hand("hand", "Pan", "H")
}
data class BoardElement(val id: String, val type: String, val color: String, val width: Double, val points: List<Point>)
data class BoardMessage(
    val type: String, val element: BoardElement? = null, val ids: List<String> = emptyList(),
    val elements: List<BoardElement> = emptyList(), val removed: List<String> = emptyList(),
)

class BoardDocument {
    private var objects = mapOf<String, BoardElement>()
    private var removed = setOf<String>()
    val elements get() = objects.values.sortedBy { it.id }
    fun snapshot() = BoardMessage("snapshot", elements = elements, removed = removed.sorted())
    fun reset() { objects = emptyMap(); removed = emptySet() }
    fun apply(message: BoardMessage) {
        if (message.type == "preview") return
        val nextObjects = objects.toMutableMap()
        val nextRemoved = removed.toMutableSet()
        val deletes = when (message.type) { "remove" -> message.ids; "snapshot" -> message.removed; else -> emptyList() }
        deletes.forEach { nextRemoved.add(it); nextObjects.remove(it) }
        val adds = when (message.type) { "put" -> listOf(requireNotNull(message.element)); "snapshot" -> message.elements; else -> emptyList() }
        adds.forEach { if (it.id !in nextRemoved && it.id !in nextObjects) nextObjects[it.id] = it }
        require(nextObjects.size <= Wire.MAX_ELEMENTS && nextRemoved.size <= Wire.MAX_REMOVED) { "This prototype board is full. Start a new board." }
        val snapshot = BoardMessage("snapshot", elements = nextObjects.values.toList(), removed = nextRemoved.toList())
        require(Wire.encode(snapshot).encodeToByteArray().size <= Wire.MAX_MESSAGE) { "This prototype board is full. Start a new board." }
        objects = nextObjects
        removed = nextRemoved
    }
}

fun constrained(start: Point, end: Point, type: String): Point {
    val dx = end.x - start.x; val dy = end.y - start.y
    if (type == "line") {
        val angle = round(atan2(dy, dx) / (PI / 4)) * (PI / 4)
        return Point(start.x + cos(angle) * hypot(dx, dy), start.y + sin(angle) * hypot(dx, dy))
    }
    val side = max(abs(dx), abs(dy))
    return Point(start.x + (if (dx < 0) -side else side), start.y + (if (dy < 0) -side else side))
}

fun outline(element: BoardElement): List<Point> {
    if (element.type == "pen" || element.type == "line") return element.points
    val a = element.points.first(); val b = element.points.last()
    val x = min(a.x, b.x); val y = min(a.y, b.y); val w = abs(b.x - a.x); val h = abs(b.y - a.y)
    if (element.type == "rectangle") return listOf(Point(x, y), Point(x + w, y), Point(x + w, y + h), Point(x, y + h), Point(x, y))
    return (0..96).map { i -> val angle = i / 96.0 * PI * 2; Point(x + w / 2 + cos(angle) * w / 2, y + h / 2 + sin(angle) * h / 2) }
}

private fun distance(p: Point, a: Point, b: Point): Double {
    val dx = b.x - a.x; val dy = b.y - a.y; val length = dx * dx + dy * dy
    val t = if (length == 0.0) 0.0 else (((p.x - a.x) * dx + (p.y - a.y) * dy) / length).coerceIn(0.0, 1.0)
    return hypot(p.x - a.x - t * dx, p.y - a.y - t * dy)
}
private fun cross(a: Point, b: Point, c: Point) = (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x)
fun hits(element: BoardElement, from: Point, to: Point, radius: Double): Boolean {
    val points = outline(element); val threshold = radius + element.width / 2
    if (points.size == 1) return distance(points[0], from, to) <= threshold
    return points.zipWithNext().any { (a, b) ->
        val crossing = cross(from, to, a) * cross(from, to, b) < 0 && cross(a, b, from) * cross(a, b, to) < 0
        crossing || minOf(distance(a, from, to), distance(b, from, to), distance(from, a, b), distance(to, a, b)) <= threshold
    }
}
